src/atomate2/vasp/sets/core.py

```diff
@@ -551,31 +402,12 @@ def __post_init__(self) -> None:
         if self.mode not in supported_modes:
             raise ValueError(f"Supported modes are: {', '.join(supported_modes)}")
 
-    def get_kpoints_updates(
-        self,
-        structure: Structure,
-        prev_incar: dict = None,
-        bandgap: float = 0.0,
-        vasprun: Vasprun = None,
-        outcar: Outcar = None,
-    ) -> dict:
+    @property
+    def kpoints_updates(self) -> dict | Kpoints:
```

Comment:
why `dict | Kpoints` and not just `dict`? if it's for parent class compat, prob the parent class needs to change

Reply:
Being able to return a Kpoints object directly is a nice feature! I think this might be in the pymatgen sets but worth checking.

Reply:
i strongly recommend homogeneous return type. just asking for trouble calling a `Kpoints` method on `kpoints_updates()` that fails because it's a `dict`.
i'd advise to always return either `dict` or `Kpoint`. given `incar_updates` is `dict`, would minimize user surprise to return `dict` here as well

Reply:
Just noting that this is used in some places in pymatgen already:
- https://github.com/materialsproject/pymatgen/blob/442d740eb7794d360cb35f187e271b3debb7f047/pymatgen/io/vasp/sets.py#L2159
- https://github.com/materialsproject/pymatgen/blob/442d740eb7794d360cb35f187e271b3debb7f047/pymatgen/io/vasp/sets.py#L2218

I think having flexibility can be useful. That said, it wouldn't be much work to allow specifying a kpoints file within a dictionary format, but it does seem like extra boilerplate.

>  given incar_updates is dict, would minimize user surprise to return dict here as well

`Incar` is already a dict so you could conceivably return `Incar` too.